consensus/core/src/ancestor.rs

```diff
@@ -0,0 +1,379 @@
+        self.state_map
+            .iter()
+            .map(|(&id, info)| (id, info.state))
+            .collect()
+    }
+
+    /// Updates the state of all ancestors based on the latest scores and quorum rounds
+    pub(crate) fn update_all_ancestors_state(&mut self) {
+        // If round prober has not run yet and we don't have network quorum round,
+        // it is okay because network_low_quorum_round will be zero and we will
+        // include all ancestors until we get more information.
+        let network_low_quorum_round = self.calculate_network_low_quorum_round();
+        let propagation_scores_by_authority = self
+            .propagation_scores
+            .scores_per_authority
+            .clone()
+            .into_iter()
+            .enumerate()
+            .map(|(idx, score)| {
+                (
+                    self.context
+                        .committee
+                        .to_authority_index(idx)
+                        .expect("Index should be valid"),
+                    score,
+                )
+            })
+            .collect::<Vec<_>>();
+
+        // If propagation scores are not ready because the first 300 commits have not
+        // happened, this is okay as we will only start excluding ancestors after that
+        // point in time.
+        for (authority_id, score) in propagation_scores_by_authority {
+            let (authority_low_quorum_round, _high) = self.quorum_round_per_authority[authority_id];
+
+            self.update_state(
+                authority_id,
+                score,
+                authority_low_quorum_round,
+                network_low_quorum_round,
+            );
+        }
+    }
+
+    /// Updates the state of the given authority based on current scores and quorum rounds.
+    fn update_state(
+        &mut self,
+        authority_id: AuthorityIndex,
+        propagation_score: u64,
+        authority_low_quorum_round: u32,
+        network_low_quorum_round: u32,
+    ) {
+        let block_hostname = &self.context.committee.authority(authority_id).hostname;
+        let ancestor_info = self.state_map.get_mut(&authority_id).unwrap_or_else(|| {
+            panic!("Expected authority_id {authority_id} to be initialized in state_map")
+        });
+
+        if ancestor_info.is_locked(
+            self.propagation_score_update_count,
+            self.quorum_round_update_count,
+        ) {
+            // If still locked, we won't make any state changes.
+            return;
+        } else {
+            ancestor_info.reset_lock();
+        }
+
+        let low_score_threshold =
+            (self.propagation_scores.highest_score() * Self::EXCLUSION_THRESHOLD_PERCENTAGE) / 100;
+
+        match ancestor_info.state {
+            // Check conditions to switch to EXCLUDE state
+            AncestorState::Include => {
+                if propagation_score <= low_score_threshold {
```

Comment:
Have you considered / tried exclusion based on the gap between a peer's low and high quorum rounds? Maybe we can leave a TODO here.

Reply:
I did consider this but thought excluding based on the quality (votes) of blocks distributed was also important but I can give this a try. Do you have a sense for a good threshold between low and high quorum rounds. I believe it should not be more than 1 round in the healthy cases but wanted to confirm. Maybe something like 5-10 rounds is a good threshold for exclusion?